src/driver_zwave.cpp

```diff
@@ -0,0 +1,452 @@
+  gwZipconnection_ = ZipConnect(serverIP_.c_str());
+
+  if (gwZipconnection_) return true;
+  return false;
+}
+
+zconnection* ZWaveDriver::ZipConnect(const char* remote_addr) {
+  if (cfgPskLen_ == 0) {
+    std::cerr << "PSK not configured - unable to connect to " << remote_addr
+              << std::endl;
+    return 0;
+  }
+
+  zconnection* zc;
+
+  zc = zclient_start(remote_addr, 41230, reinterpret_cast<char*>(&cfgPsk_[0]),
+                     cfgPskLen_, ApplicationCommandHandler);
+  if (zc == 0) {
+    std::cout << "Error connecting to " << remote_addr << std::endl;
+  } else {
+    std::cout << "ZWaveDriver connected to " << remote_addr << std::endl;
+  }
+  return zc;
+}
+
+void print_hex_string(const uint8_t* data, unsigned int datalen) {
+  unsigned int i;
+
+  for (i = 0; i < datalen; i++) {
+    std::cout << " " << std::hex << int(data[i]);
+    if ((i & 0xf) == 0xf) {
+      std::cout << std::endl;
+    }
+  }
+  std::cout.flush();
+}
+
+void ZWaveDriver::ApplicationCommandHandler(zconnection* /*zc*/,
+                                            const uint8_t* data,
+                                            uint16_t datalen) {
+  int i;
+  int len;
+  const uint8_t COMMAND_CLASS_NETWORK_MANAGEMENT_INCLUSION = 0x34;
+
+  unsigned char cmd_classes[400][MAX_LEN_CMD_CLASS_NAME];
+  std::cout << "ApplicationCommandHandler datalen=" << datalen << std::endl;
+
+  print_hex_string(data, datalen);
+
+  switch (data[0]) {
+    case COMMAND_CLASS_NETWORK_MANAGEMENT_INCLUSION:
+      parse_network_mgmt_inclusion_packet(data, datalen);
+      break;
+
+    default:
+      memset(cmd_classes, 0, sizeof(cmd_classes));
+      /* decode() clobbers data - but we are not using it afterwards, hence
+       * the
+       * typecast */
+      decode((uint8_t*)data, datalen, cmd_classes, &len);
+      std::cout << std::endl;
+      for (i = 0; i < len; i++) {
+        std::cout << " +++ " << cmd_classes[i] << std::endl;
+      }
+      std::cout << std::endl;
+      break;
+  }
+
+  ZwaveMsg msg;
+
+  if (datalen >= 3 && data[2])
+    msg.set_result(true);
+  else
+    msg.set_result(false);
+
+  std::string buffer;
+  msg.SerializeToString(&buffer);
+  static_zqm_push_update_->Send(buffer);
+}
+
+void ZWaveDriver::TransmitDonePan(zconnection* /*zc*/,
+                                  transmission_status_code_t status) {
+  std::cout << "ZWaveDriver::transmit_done_pan" << std::endl;
+
+  switch (status) {
+    case TRANSMIT_OK:
+      break;
+    case TRANSMIT_NOT_OK:
+      std::cerr << "Transmit failed" << std::endl;
+      break;
+    case TRANSMIT_TIMEOUT:
+      std::cerr << "Transmit attempt timed out" << std::endl;
+      break;
+  }
+  ZWaveDriver::panConnectionBusy_ = false;
+}
+
+static int hex2int(char c) {
+  if (c >= '0' && c <= '9') {
+    return c - '0';
+  } else if (c >= 'a' && c <= 'f') {
+    return c - 'a' + 0xa;
+  } else if (c >= 'A' && c <= 'F') {
+    return c - 'A' + 0xa;
+  } else {
+    return -1;
```

Comment:
It seems that this can happen when c is 0. Perhaps this should be handled outside?